src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs

```diff
@@ -113,6 +113,7 @@ public override string SubProtocol
                         secureConnection ? Interop.WinHttp.WINHTTP_FLAG_SECURE : 0);
 
                     ThrowOnInvalidHandle(_operation.RequestHandle);
+                    _operation.IncrementHandlesOpenWithCallback();
```

Comment:
`WinHttpWebSocketState.Unpin` has a check in it so that it won't do anything unless the GCHandle has been allocated, i.e. `Pin` has been called already.

The only way callbacks can happen in general on these handles is if the callback has been fully wired up with the callback and the context value. That is why the Pin() call is right after the final call to wire up the callback.

In terms of anything bad going wrong between the `_operation.IncrementHandlesOpenWithCallback' and the `Pin` call, we don't really care.  There is lots of code between those calls. But the callback hasn't been wired up yet. So, there won't be a need to protect the state object (i.e. pin/strong root it) because it will never be dereferenced thru a native callback.